### ISSUE TITLE
vSphere: DAX not supported by filesystem

### ISSUE THREAD

**Issue by Tianyang-Zhang** [intel/pmem-csi#884]
Hi folks, I'm trying to use PMEM-CSI with a 2 node OpenShift 4.6.9 cluster. I wasn't able to install PMEM-CSI using the operator. It seems caused by the OpenShift privilege issue mentioned in https://github.com/intel/pmem-csi/issues/652. 

Then, I installed PMEM-CSI from the source(`devel` branch) by adding the OpenShift special permission to deploy/kubernetes-1.20/pmem-csi-lvm.yaml(mentioned in the example in #652). 

During the installation, I cannot setup the certificates by following the install.md:
```
$ curl -L https://pkg.cfssl.org/R1.2/cfssl_linux-amd64 -o _work/bin/cfssl --create-dirs
$ curl -L https://pkg.cfssl.org/R1.2/cfssljson_linux-amd64 -o _work/bin/cfssljson --create-dirs
$ chmod a+x _work/bin/cfssl _work/bin/cfssljson
$ KUBCONFIG="<<your cluster kubeconfig path>>" PATH="$PWD/_work/bin:$PATH" ./test/setup-ca-kubernetes.sh
```
The `setup-ca-kubernetes.sh` returns following error because the `.pem` certificate file does not exists:
```
[root@aep0 pmem-csi]# ./test/setup-ca-kubernetes.sh
realpath: /home/tomz/container/openshift/pmem-csi/pmem-csi/_work/pmem-ca/ca.pem: No such file or directory
realpath: /home/tomz/container/openshift/pmem-csi/pmem-csi/_work/pmem-ca/ca-key.pem: No such file or directory
Generating Certificate for 'pmem-registry'(NS=pmem-csi) ...
{
    "CN": "pmem-registry",
    "hosts": [
        "127.0.0.1",
"pmem-csi-scheduler", "pmem-csi-scheduler.pmem-csi", "pmem-csi-scheduler.pmem-csi.svc",
"pmem-csi-intel-com-scheduler", "pmem-csi-intel-com-scheduler.pmem-csi", "pmem-csi-intel-com-scheduler.pmem-csi.svc",
"pmem-csi-metrics", "pmem-csi-metrics.pmem-csi", "pmem-csi-metrics.pmem-csi.svc",
"pmem-csi-intel-com-metrics", "pmem-csi-intel-com-metrics.pmem-csi", "pmem-csi-intel-com-metrics.pmem-csi.svc",
        "pmem-registry"
    ],
    "key": {
        "algo": "ecdsa",
        "size": 256
    }
}
2021/02/04 16:19:11 [ERROR] need a CA certificate (provide one with -ca)
Failed to parse input: unexpected end of JSON input
Generating Certificate for 'pmem-node-controller'(NS=pmem-csi) ...
{
    "CN": "pmem-node-controller",
    "hosts": [

        "pmem-node-controller"
    ],
    "key": {
        "algo": "ecdsa",
        "size": 256
    }
}
2021/02/04 16:19:11 [ERROR] need a CA certificate (provide one with -ca)
Failed to parse input: unexpected end of JSON input
```
The `_work` directory only contains `bin`, I couldn't find any resource to get `pmem-ca` from the installation guide.
```
ls -l _work/
total 0
drwxr-x--- 2 root root 36 Feb  3 15:38 bin
```

So I comment out the StatefulSet `pmem-csi-intel-com-controller-0` in `pmem-csi-lvm.yaml`.

Before the deployment, I use `sudo vgcreate --force ndbus0region0fsdax /dev/pmem0 && sudo vgchange --addtag fsdax ndbus0region0fsdax` to create VG since I deployed the OpenShift cluster on vSphere and ndctl doesn't support label storage area in ESXi. This is also mentioned in the PMEM-CSI's GCE example: https://github.com/intel/pmem-csi/blob/devel/examples/gce.md#installing-pmem-csi.
`vgdisplay` output on the CoreOS worker nodes:
```
$ sudo vgdisplay
  --- Volume group ---
  VG Name               ndbus0region0fsdax
  System ID
  Format                lvm2
  Metadata Areas        1
  Metadata Sequence No  10
  VG Access             read/write
  VG Status             resizable
  MAX LV                0
  Cur LV                0
  Open LV               0
  Max PV                0
  Cur PV                1
  Act PV                1
  VG Size               <50.00 GiB
  PE Size               4.00 MiB
  Total PE              12799
  Alloc PE / Size       0 / 0
  Free  PE / Size       12799 / <50.00 GiB
  VG UUID               drv3Pg-2cFx-IFMD-YY2h-kib3-BsyJ-LPxSG1
```

Then, I labeled all worker nodes and deployed PMEM-CSI and the following pods are running normally:
```
$ oc get nodes --show-labels
...
osc-nkmp7-worker-cn5f5   Ready    worker   47h   v1.19.0+7070803   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=osc-nkmp7-worker-cn5f5,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhcos,pmem-csi.intel.com/node=osc-nkmp7-worker-cn5f5,storage=pmem
osc-nkmp7-worker-phzst   Ready    worker   46h   v1.19.0+7070803   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=osc-nkmp7-worker-phzst,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhcos,pmem-csi.intel.com/node=osc-nkmp7-worker-phzst,storage=pmem
...

$ oc create -f kubernetes-1.20/pmem-csi-lvm.yaml
$ oc get pods -n pmem-csi
NAME                            READY   STATUS    RESTARTS   AGE
pmem-csi-intel-com-node-qpghw     3/3     Running   0          21h
pmem-csi-intel-com-node-zkksc     3/3     Running   0          21h
```

Create SC and PVC:
```
$ oc create -f https://github.com/intel/pmem-csi/raw/devel/deploy/common/pmem-storageclass-ext4.yaml
$ oc create -f https://github.com/intel/pmem-csi/raw/devel/deploy/common/pmem-storageclass-xfs.yaml
$ oc get sc
NAME               PROVISIONER                    RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
pmem-csi-sc-ext4   pmem-csi.intel.com             Delete          Immediate           false                  26h
pmem-csi-sc-xfs    pmem-csi.intel.com             Delete          Immediate           false                  26h
thin (default)     kubernetes.io/vsphere-volume   Delete          Immediate           false                  47h

$ oc create -f ./pvc-hazelcast.yaml
$ oc get pvc
NAME                         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
pmem-csi-pvc-xfs-hazelcast   Bound    pvc-4cafe87a-a435-43d7-806d-bcc5f7c99ebe   20Gi       RWO            pmem-csi-sc-xfs   21h
```

Create application pod with node affinity spec:
```
...
spec:
  affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: storage
                operator: In
                values:
                - pmem
...
```
but I can only see one of the worker nodes has `/dev/ndbus0region0fsdax/` device after the application pod starts. I attached to my application containers and found that `pmem_map_file()` returns `not PMEM` although I can see the device mount through `df`. It seems my application pod is scheduled on the other node.

Do you have any clue what I missed? Or is this related to the controller StatefulSet? Thanks!

**Comment by pohly** [comment]
Thanks for the detailed report. @avalluri : can you look into this?

**Comment by avalluri** [comment]
> The setup-ca-kubernetes.sh returns following error because the .pem certificate file does not exists:

The certificate/secret creation scripts are more centric to out test cluster. #885 should fix this.

**Comment by pohly** [comment]
> The certificate/secret creation scripts are more centric to out test cluster.

That depends. We still need them for creating certificates in advance for the scheduler extensions. Users can also do that differently of course, but those scripts are what we document.

@avalluri have you also looked into the issue of starting pods on the right nodes?

**Comment by avalluri** [comment]
> @avalluri have you also looked into the issue of starting pods on the right nodes?

Not really, I doubted the reason could be that the scheduler extension(controller) was not running that might have led to pick the wrong node.  Though. other possible reason could be the manual LVM volume group creation is missing in one node, but i presume not the case.

**Comment by Tianyang-Zhang** [comment]
Thanks! I will try this fix and inform you the update.

**Comment by pohly** [comment]
@Tianyang-Zhang : regarding bringing up the statefulset pods on each node - are you sure that LVM was set up on all nodes that you have labeled for the driver?

And another, unrelated question: are you working for Intel? In that case we are also available on internal chat, if you prefer that.

**Comment by Tianyang-Zhang** [comment]
Hi @pohly.  I just applied the fix from @avalluri. The controller successfully started, but the PMEM provision issue still exist.

Yes, the LVM is set on two worker nodes. Here are the two worker nodes:
```
osc-nkmp7-worker-cn5f5   Ready    worker   5d18h   v1.19.0+7070803   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=osc-nkmp7-worker-cn5f5,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhcos,pmem-csi.intel.com/node=osc-nkmp7-worker-cn5f5,storage=pmem
osc-nkmp7-worker-phzst   Ready    worker   5d16h   v1.19.0+7070803   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=osc-nkmp7-worker-phzst,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhcos,pmem-csi.intel.com/node=osc-nkmp7-worker-phzst,storage=pmem
```

Here is the `vgdisplay` output of both nodes:
```
# node cn5f5
[core@osc-nkmp7-worker-cn5f5 ~]$ sudo vgdisplay
  --- Volume group ---
  VG Name               ndbus0region0fsdax
  System ID
  Format                lvm2
  Metadata Areas        1
  Metadata Sequence No  12
  VG Access             read/write
  VG Status             resizable
  MAX LV                0
  Cur LV                1
  Open LV               0
  Max PV                0
  Cur PV                1
  Act PV                1
  VG Size               <50.00 GiB
  PE Size               4.00 MiB
  Total PE              12799
  Alloc PE / Size       5120 / 20.00 GiB
  Free  PE / Size       7679 / <30.00 GiB
  VG UUID               drv3Pg-2cFx-IFMD-YY2h-kib3-BsyJ-LPxSG1

# node phzst
[core@osc-nkmp7-worker-phzst ~]$ sudo vgdisplay
  --- Volume group ---
  VG Name               ndbus0region0fsdax
  System ID
  Format                lvm2
  Metadata Areas        1
  Metadata Sequence No  25
  VG Access             read/write
  VG Status             resizable
  MAX LV                0
  Cur LV                0
  Open LV               0
  Max PV                0
  Cur PV                1
  Act PV                1
  VG Size               <50.00 GiB
  PE Size               4.00 MiB
  Total PE              12799
  Alloc PE / Size       0 / 0
  Free  PE / Size       12799 / <50.00 GiB
  VG UUID               9ofjI0-Qu7e-bbNy-nx1G-iT1v-KOx3-7mdiux

```
As you can see above, the `cn5f5` node has Alloc PE size 20.00 GB, which matches what I have in the PVC yaml file. But for  `phzst` node, the allocated PE is 0.

Unfortunately, I'm not working for Intel. Please let me know if you prefer any other chat, I can also show you the entire flow I did via screen-share/video if you need. Thanks!

**Comment by pohly** [comment]
> As you can see above, the cn5f5 node has Alloc PE size 20.00 GB, which matches what I have in the PVC yaml file.

That matches what you quoted earlier about one volume being created:
```
$ oc get pvc
NAME                         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
pmem-csi-pvc-xfs-hazelcast   Bound    pvc-f5c8fc7d-93a4-49bb-9f46-9e9d72390e06   20Gi       RWO            pmem-csi-sc-xfs   14m
```

If you look at the full output for volumes (`kubectl get -o yaml pv`), then it should show that there is one which has `cn5f5` as topology. 

> But for phzst node, the allocated PE is 0.

That looks fine to me. Volumes only get created on one node.

> Additionally, I got these controller log:
>  ... pvc-62ee033f-b43e-43f9-a300-ec17767e9d8d ...  'Warning' reason: 'VolumeFailedDelete' not implemented

This is another persistent volume. The error looks a bit odd.

On which node does pvc-62ee033f-b43e-43f9-a300-ec17767e9d8d exist (topology in `-o yaml` output) and is this error from the provisioner running on that node or some other node?

> Unfortunately, I'm not working for Intel. Please let me know if you prefer any other chat, I can also show you the entire flow I did via screen-share/video if you need. Thanks!

I was just wondering. Let's stick with remote debugging for now and we switch to a life session if we get stuck.

**Comment by Tianyang-Zhang** [comment]
>>Additionally, I got these controller log:
... pvc-62ee033f-b43e-43f9-a300-ec17767e9d8d ... 'Warning' reason: 'VolumeFailedDelete' not implemented

>This is another persistent volume. The error looks a bit odd.

Sorry, I forgot to mention that I reset the environment when applying the CA fix. I just updated the content in the issue description. The PVC I have right now is:
```
$ oc get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                  STORAGECLASS      REASON   AGE
pvc-4cafe87a-a435-43d7-806d-bcc5f7c99ebe   20Gi       RWO            Delete           Bound    hazelcast/pmem-csi-pvc-xfs-hazelcast   pmem-csi-sc-xfs            21h
```

But I cannot find where is `pvc-62ee033f-b43e-43f9-a300-ec17767e9d8d` exist. Here is the out put of `kubectl get pv` and the full output of `kubectl get -o yaml pv`, I can only see `pvc-4cafe87a-a435-43d7-806d-bcc5f7c99ebe`:
```
$ kubectl get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                  STORAGECLASS      REASON   AGE
pvc-4cafe87a-a435-43d7-806d-bcc5f7c99ebe   20Gi       RWO            Delete           Bound    hazelcast/pmem-csi-pvc-xfs-hazelcast   pmem-csi-sc-xfs            21h
```

```
$ kubectl get -o yaml pv
apiVersion: v1
items:
- apiVersion: v1
  kind: PersistentVolume
  metadata:
    annotations:
      pv.kubernetes.io/provisioned-by: pmem-csi.intel.com
    creationTimestamp: "2021-02-08T19:52:45Z"
    finalizers:
    - kubernetes.io/pv-protection
    managedFields:
    - apiVersion: v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:pv.kubernetes.io/provisioned-by: {}
        f:spec:
          f:accessModes: {}
          f:capacity:
            .: {}
            f:storage: {}
          f:claimRef:
            .: {}
            f:apiVersion: {}
            f:kind: {}
            f:name: {}
            f:namespace: {}
            f:resourceVersion: {}
            f:uid: {}
          f:csi:
            .: {}
            f:driver: {}
            f:fsType: {}
            f:volumeAttributes:
              .: {}
              f:storage.kubernetes.io/csiProvisionerIdentity: {}
            f:volumeHandle: {}
          f:nodeAffinity:
            .: {}
            f:required:
              .: {}
              f:nodeSelectorTerms: {}
          f:persistentVolumeReclaimPolicy: {}
          f:storageClassName: {}
          f:volumeMode: {}
      manager: csi-provisioner
      operation: Update
      time: "2021-02-08T19:52:45Z"
    - apiVersion: v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:phase: {}
      manager: kube-controller-manager
      operation: Update
      time: "2021-02-08T19:52:45Z"
    name: pvc-4cafe87a-a435-43d7-806d-bcc5f7c99ebe
    resourceVersion: "2295680"
    selfLink: /api/v1/persistentvolumes/pvc-4cafe87a-a435-43d7-806d-bcc5f7c99ebe
    uid: d774ea9e-3ac2-4f51-a018-9f4949a0ade3
  spec:
    accessModes:
    - ReadWriteOnce
    capacity:
      storage: 20Gi
    claimRef:
      apiVersion: v1
      kind: PersistentVolumeClaim
      name: pmem-csi-pvc-xfs-hazelcast
      namespace: hazelcast
      resourceVersion: "2295676"
      uid: 4cafe87a-a435-43d7-806d-bcc5f7c99ebe
    csi:
      driver: pmem-csi.intel.com
      fsType: xfs
      volumeAttributes:
        storage.kubernetes.io/csiProvisionerIdentity: 1612813238744-8081-pmem-csi.intel.com-osc-nkmp7-worker-cn5f5
      volumeHandle: pvc-4c-a929f81c172dbc4ffda9937437e400e68c3643f60f7c3cafc7505f31
    nodeAffinity:
      required:
        nodeSelectorTerms:
        - matchExpressions:
          - key: pmem-csi.intel.com/node
            operator: In
            values:
            - osc-nkmp7-worker-cn5f5
    persistentVolumeReclaimPolicy: Delete
    storageClassName: pmem-csi-sc-xfs
    volumeMode: Filesystem
  status:
    phase: Bound
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

**Comment by pohly** [comment]
That all looks pretty normal.

> But I cannot find where is pvc-62ee033f-b43e-43f9-a300-ec17767e9d8d exist.

That's shown here:
```
$ kubectl get -o yaml pv
apiVersion: v1
items:
- apiVersion: v1
...
    nodeAffinity:
      required:
        nodeSelectorTerms:
        - matchExpressions:
          - key: pmem-csi.intel.com/node
            operator: In
            values:
            - osc-nkmp7-worker-cn5f5
...
```

**Comment by Tianyang-Zhang** [comment]
> That all looks pretty normal.
> 
> > But I cannot find where is pvc-62ee033f-b43e-43f9-a300-ec17767e9d8d exist.
> 
> That's shown here:
> 
> ```
> $ kubectl get -o yaml pv
> apiVersion: v1
> items:
> - apiVersion: v1
> ...
>     nodeAffinity:
>       required:
>         nodeSelectorTerms:
>         - matchExpressions:
>           - key: pmem-csi.intel.com/node
>             operator: In
>             values:
>             - osc-nkmp7-worker-cn5f5
> ...
> ```

I think this node has the `pvc-4cafe87a-a435-43d7-806d-bcc5f7c99ebe` provisioned but not `pvc-62ee033f-b43e-43f9-a300-ec17767e9d8d,  isn't it?

Maybe I should try to remove the PMEM label on the node `phzst` and try if everything works well?

**Comment by pohly** [comment]
Now I see, we are talking about two different PVs. I missed that - these UUIDs are a bit hard to read...

The `get pv` output only shows one. `pvc-...bcc5f7c99ebe` doesn't exist anymore but somehow PVC `pmem-csi-pvc-xfs-hazelcast` is bound to it. How this happened isn't clear.

My suggestion is to force-delete all PVs and PVCs (see https://github.com/kubernetes-csi/external-provisioner/#deleting-local-volumes-after-a-node-failure-or-removal for PV - the same works for PVCs, if needed). Then also delete all logical LVM volumes that you have on one node. That restores a clean state. Kill driver pods to ensure that they also have no hidden state, but that shouldn't be necessary.

Then start again by creating one PVC and checking that this works. Are you using late binding or immediate binding in your storage class? Both should work, but late binding is recommended - we should have that documented somewhere, but I currently don't find it. I'll enhance that in the documentation.

BTW, thanks for trying out our master branch. Using the latest stable release might have been easier because you are running into issues with our recent changes, but we do appreciate the feedback before tagging 0.9.0.

**Comment by pohly** [comment]
> Maybe I should try to remove the PMEM label on the node phzst and try if everything works well?

It should work with two nodes.

**Comment by Tianyang-Zhang** [comment]
I tried the method you suggests several times but it still not working. I notice that there is a warning message when I creating the certificate, but it seems harmless.
```
2021/02/12 12:05:15 [WARNING] This certificate lacks a "hosts" field. This makes it unsuitable for
websites. For more information see the Baseline Requirements for the Issuance and Management
of Publicly-Trusted Certificates, v.1.1.6, from the CA/Browser Forum (https://cabforum.org);
```

I will re-install the whole OpenShift cluster to make sure it is fresh and no hidden states. Thanks for following up, I will update after the re-installation.

**Comment by Tianyang-Zhang** [comment]
Hi @pohly, I re-created the OpenShift cluster to make sure everything is fresh. Unfortunately, I still got the same `is not PMEM` error. The storage class I create is using late binding:
```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: pmem-csi-sc-xfs
parameters:
  csi.storage.k8s.io/fstype: xfs
  eraseafter: "false"
provisioner: pmem-csi.intel.com
volumeBindingMode: WaitForFirstConsumer
```

I see you mentioned using `allowedTopologies` in the installation guide. Could you please provide more details of applying this flag? Thanks!

**Comment by pohly** [comment]
> I attached to my application containers and found that pmem_map_file() returns not PMEM although I can see the device mount through df. It seems my application pod is scheduled on the other node.

I think this is where this went a bit astray: we ended up debugging the "scheduled on the other node" part and ignored the possibility that everything had been set up as expected, but then failed to provide a volume with DAX support.

Can you explain where the "not PMEM" error comes from? Does the application get that from the Linux kernel?

One way to test that is to run the `pmem-csi-driver-test` image itself as the application with "sleep infinity" as the command, with a PMEM volume mounted somewhere. Then log into that container and run `/usr/local/bin/pmem-dax-check /<PMEM-VOLUME>/test-file`.

The source code of that command is here:
https://github.com/intel/pmem-csi/blob/devel/test/cmd/pmem-dax-check/main.go

It does the `mmap(MAP_SHARED_VALIDATE|MAP_SYNC)` and will tell us whether the kernel supports it.

If that fails, then we need to double-check that the volume was mounted with `-odax`. If that is the case, then we'll have to consult with our kernel developers because I am running out of ideas. I thought that once the filesystem accepts a mount with `-odax` that then `MAP_SYNC` is supported for all files.

What is the version of the kernel?

**Comment by pohly** [comment]
> I see you mentioned using allowedTopologies in the installation guide. Could you please provide more details of applying this flag? 

Please ignore that. That sentence shouldn't have been merged, it was for an experiment that didn't work out.

**Comment by Tianyang-Zhang** [comment]
The "not PMEM" error I got is from the `pmem_map_file()` function call of `libpmem`. The `int *is_pmemp` arg returns 0 which indicates the device is not PMEM.

The OpenShift 4.6.9 vShpere IPI deploys the worker nodes in CoreOS 4.6, with 4.18.0 kernel. Here are the details:
```
[core@osc-szm5p-worker-ql6x9 ~]$ cat /proc/version
Linux version 4.18.0-193.37.1.el8_2.x86_64 (mockbuild@x86-vm-09.build.eng.bos.redhat.com) (gcc version 8.3.1 20191121 (Red Hat 8.3.1-5) (GCC)) #1 SMP Sun Dec 6 19:59:00 EST 2020

[core@osc-szm5p-worker-ql6x9 ~]$ cat /etc/os-release
NAME="Red Hat Enterprise Linux CoreOS"
VERSION="46.82.202012151054-0"
VERSION_ID="4.6"
OPENSHIFT_VERSION="4.6"
RHEL_VERSION="8.2"
PRETTY_NAME="Red Hat Enterprise Linux CoreOS 46.82.202012151054-0 (Ootpa)"
ID="rhcos"
ID_LIKE="rhel fedora"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:8::coreos"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="OpenShift Container Platform"
REDHAT_BUGZILLA_PRODUCT_VERSION="4.6"
REDHAT_SUPPORT_PRODUCT="OpenShift Container Platform"
REDHAT_SUPPORT_PRODUCT_VERSION="4.6"
OSTREE_VERSION='46.82.202012151054-0'
```

Below are the outputs I got from `pmem-csi-driver-test` image and `/usr/local/bin/pmem-dax-check`. I created a PVC, mount it to `/mnt/pmem/`, and ran a init container to fallocate a file in that directory, then attached to `pmem-csi-driver-test` container to run the test program.  Eventually, I got `does not support MAP_SYNC: operation not supported` error from `/usr/local/bin/pmem-dax-check`. I'm not sure if these steps are the same as you suggests.
```
$ oc get sc
NAME              PROVISIONER                    RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
pmem-csi-sc-xfs   pmem-csi.intel.com             Delete          WaitForFirstConsumer   false                  22h

$ oc get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                STORAGECLASS      REASON   AGE
pvc-a65db8ac-56a6-4a08-b629-f8f4c7026d02   20Gi       RWO            Delete           Bound    default/pmem-csi-pvc-xfs-hazelcast   pmem-csi-sc-xfs            25m

$ oc get pvc
NAME                         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
pmem-csi-pvc-xfs-hazelcast   Bound    pvc-a65db8ac-56a6-4a08-b629-f8f4c7026d02   20Gi       RWO            pmem-csi-sc-xfs   12m
```

```
pmem-csi@test-pmem-csi:/$ df -h
Filesystem                                                                               Size  Used Avail Use% Mounted on
overlay                                                                                  120G   12G  109G  10% /
tmpfs                                                                                     64M     0   64M   0% /dev
tmpfs                                                                                    3.9G     0  3.9G   0% /sys/fs/cgroup
shm                                                                                       64M     0   64M   0% /dev/shm
tmpfs                                                                                    3.9G   46M  3.9G   2% /etc/hostname
/dev/ndbus0region0fsdax/pvc-a6-e11e8fc4774109ee926ced273cca808beb924d51fb4ecf6e490bb055   20G   16G  5.0G  76% /mnt/pmem
/dev/mapper/coreos-luks-root-nocrypt                                                     120G   12G  109G  10% /etc/hosts
tmpfs                                                                                    3.9G   28K  3.9G   1% /run/secrets/kubernetes.io/serviceaccount
tmpfs                                                                                    3.9G     0  3.9G   0% /proc/acpi
tmpfs                                                                                    3.9G     0  3.9G   0% /proc/scsi
tmpfs                                                                                    3.9G     0  3.9G   0% /sys/firmware

pmem-csi@test-pmem-csi:/$ ls -lh /mnt/pmem/
total 15G
-rw-r--r--. 1 root root 15G Feb 17 22:58 mm

pmem-csi@test-pmem-csi:/$ /usr/local/bin/pmem-dax-check /mnt/pmem/mm
/mnt/pmem/mm: does not support MAP_SYNC: operation not supported
```

And here is the description of `pmem-csi-driver-test` container:
```
Containers:
  pmem-csi-driver-testr:
    Container ID:  cri-o://fdf32ce8f9a22cb0c103b9a5adc9accc41484940886cbf237f7c468875bbdbcf
    Image:         intel/pmem-csi-driver-test:canary
    Image ID:      docker.io/intel/pmem-csi-driver-test@sha256:b330003056697f4acfaed3d6cd77f150739cb89ebf7f3251b4495810030267fc
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/bash
      -c
      sleep infinity
    State:          Running
      Started:      Wed, 17 Feb 2021 14:59:03 -0800
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /mnt/pmem from pmem-pv-storage (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-rlwtz (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  pmem-pv-storage:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  pmem-csi-pvc-xfs-hazelcast
    ReadOnly:   false
```

> I think this is where this went a bit astray: we ended up debugging the "scheduled on the other node" part and ignored the possibility that everything had been set up as expected, but then failed to provide a volume with DAX support.

I think you are right, the title of this issue is probably not proper, please change the issue title if you think it's necessary, thanks!

**Comment by okartau** [comment]
It did not become clear for me from above, does the problem of missing DAX support happen also for fstype:ext4 ?
The logs and state view mostly show it for fstype:xfs, so the question arises: is this issue xfs-specific?

**Comment by pohly** [comment]
> Eventually, I got does not support MAP_SYNC: operation not supported error from /usr/local/bin/pmem-dax-check. I'm not sure if these steps are the same as you suggests.

Yes, that's it.

Can you check with "mount" that the filesystem on /dev/ndbus0region0fsdax/pvc-a6-e11e8fc4774109ee926ced273cca808beb924d51fb4ecf6e490bb055 was mounted with `-odax`?

> It did not become clear for me from above, does the problem of missing DAX support happen also for fstype:ext4 ?
> The logs and state view mostly show it for fstype:xfs, so the question arises: is this issue xfs-specific?

It's definitely worth trying with ext4.

**Comment by Tianyang-Zhang** [comment]
I don't see the `dax` mount option(I retried and still don't see `dax`):
```
mount | grep /mnt/pmem
/dev/ndbus0region0fsdax/pvc-a6-e11e8fc4774109ee926ced273cca808beb924d51fb4ecf6e490bb055 on /mnt/pmem type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
```

When I try ext4, it seems the kernel CoreOS uses is not supporting mount ext4 with `-o dax` option:
```
$ oc get pvc
NAME                     STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS       AGE
pmem-csi-pvc-ext4-test   Bound    pvc-53018546-0168-4135-91f6-801fa7b783ba   20Gi       RWO            pmem-csi-sc-ext4   83s
```

Pod description:
```
Events:
  Type     Reason       Age               From               Message
  ----     ------       ----              ----               -------
  Normal   Scheduled    25s               default-scheduler  Successfully assigned default/test-pmem-csi to osc-szm5p-worker-ql6x9
  Warning  FailedMount  9s (x6 over 25s)  kubelet            MountVolume.MountDevice failed for volume "pvc-53018546-0168-4135-91f6-801fa7b783ba" : rpc error: code = Internal desc = mount filesystem failed: command "/bin/mount -c -o dax /dev/ndbus0region0fsdax/pvc-53-9d1d7ec2f678ae5cd0176ec6d3af236c1c5fde37dcfa6864ef032f13 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-53018546-0168-4135-91f6-801fa7b783ba/globalmount" failed: exit status 32
Output: mount: /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-53018546-0168-4135-91f6-801fa7b783ba/globalmount: wrong fs type, bad option, bad superblock on /dev/ndbus0region0fsdax/pvc-53-9d1d7ec2f678ae5cd0176ec6d3af236c1c5fde37dcfa6864ef032f13, missing codepage or helper program, or other error.
```

**Comment by pohly** [comment]
> I don't see the dax mount option when I use xfs

That's a bit surprising. I'll double check whether that should be present tomorrow.

But either way, it looks like the kernel simply doesn't support PMEM. Here are some relevant kernel config options (from a Debian kernel):
```
$ grep DAX /boot/config-4.19.0-14-amd64 
CONFIG_NVDIMM_DAX=y
CONFIG_DAX_DRIVER=y
CONFIG_DAX=y
CONFIG_DEV_DAX=m
CONFIG_DEV_DAX_PMEM=m
CONFIG_FS_DAX=y
CONFIG_FS_DAX_PMD=y
```

Can you investigate whether those are present in the kernel?

**Comment by Tianyang-Zhang** [comment]
Yes, these configs present in the kernel config:
```
[core@osc-szm5p-worker-ql6x9 ~]$ cat /lib/modules/$(uname -r)/config | grep DAX
CONFIG_NVDIMM_DAX=y
CONFIG_DAX_DRIVER=y
CONFIG_DAX=y
CONFIG_DEV_DAX=m
CONFIG_DEV_DAX_PMEM=m
CONFIG_DEV_DAX_PMEM_COMPAT=m
CONFIG_FS_DAX=y
CONFIG_FS_DAX_PMD=y
```

**Comment by pohly** [comment]
Was the underlying PMEM set up in fsdax mode? I don't remember whether we discussed this.

This section has some instructions which may or may not be applicable to vSphere:
https://software.intel.com/content/www/us/en/develop/articles/provision-intel-optane-dc-persistent-memory-for-kvm-qemu-guests.html

**Comment by pohly** [comment]
I can confirm that mounting an XFS volume silently ignores the -odax:
```
[fedora@pmem-csi-pmem-govm-worker3 ~]$ sudo ndctl create-namespace -m raw
{
  "dev":"namespace0.0",
  "mode":"raw",
  "size":"63.98 GiB (68.70 GB)",
  "uuid":"1a7da011-e570-4706-ba59-bd292c06e453",
  "sector_size":512,
  "blockdev":"pmem0"
}
[fedora@pmem-csi-pmem-govm-worker3 ~]$ ls /dev/pmem0 
/dev/pmem0
[fedora@pmem-csi-pmem-govm-worker3 ~]$ sudo mkdir /tmp/mnt
[fedora@pmem-csi-pmem-govm-worker3 ~]$ sudo mkfs.xfs -b size=4096 -m reflink=0 -f /dev/pmem0 
meta-data=/dev/pmem0             isize=512    agcount=4, agsize=4193280 blks
         =                       sectsz=4096  attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=1, rmapbt=0
         =                       reflink=0
data     =                       bsize=4096   blocks=16773120, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=8190, version=2
         =                       sectsz=4096  sunit=1 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
[fedora@pmem-csi-pmem-govm-worker3 ~]$ sudo mount -odax /dev/pmem0 /tmp/mnt
[fedora@pmem-csi-pmem-govm-worker3 ~]$ mount
...
/dev/pmem0 on /tmp/mnt type xfs (rw,relatime,seclabel,attr2,dax=never,inode64,logbufs=8,logbsize=32k,noquota)
```

I also do get the message from https://docs.pmem.io/ndctl-user-guide/troubleshooting:
```
$ dmesg
...
[ 3115.869604] XFS (pmem0): DAX enabled. Warning: EXPERIMENTAL, use at your own risk
[ 3115.870245] XFS (pmem0): DAX unsupported by block device. Turning off DAX.
```

We haven't had that issue before because normally, PMEM-CSI itself creates the namespace that it then uses for LVM, so we never run into this. I wonder whether we should add some code to detect it or just extend our documentation for those who do this initial setup manually - we should at least document it, because the manual setup is currently hard to find in the gce example.

**Comment by okartau** [comment]
> because normally, PMEM-CSI itself creates the namespace

I think by design, PMEM-CSI should always operate on namespaces created by itself, in fsdax mode. Is the namespace in current case not created by PMEM-CSI? Is it not in fsdax mode? 
but yes it is posisble to bypass "foreign" check by creating namespace manually and give it certain Name attribute,
in that sense the "foreign vs own detection" is not very secure.

> /dev/ndbus0region0fsdax/pvc-53-9d1d7ec2f678ae5cd0176ec6d3af236c1c5fde37dcfa6864ef032f13

The naming of device however leaves impression it was created by PMEM-CSI, so it has to be in fsdax mode.

> sudo ndctl create-namespace -m raw

In your example, by manually forcing mode=raw, of course it will not support DAX. But namespace created by PMEM-CSI should always come in fsdax mode, we don't have other modes supported.

**Comment by okartau** [comment]
I found related discussion and analysis in #23, where it was detected that in some combination of kernel/distro, dax capability was not propagating for LV devices. The kernel 4.18 was mentioned there BTW. So it may be here we hit the same? 
Can we cross-checking using direct mode to exclude LV-specific potential troubles?

**Comment by pohly** [comment]
> Is the namespace in current case not created by PMEM-CSI?

No, it was created manually.

**Comment by Tianyang-Zhang** [comment]
> Can we cross-check using direct mode to exclude LV-specific potential troubles?

In GCE example, it mentions that direct mode does not work on VMs since `ndctl init-label` doesn't work. I verified this with `ndctl` team that vSphere ESXi also doesn't support label storage area. Is there any workaround to run direct mode on vSphere VM without using LSA?

> > Is the namespace in current case not created by PMEM-CSI?
> 
> No, it was created manually.

In my case with OpenShift, the namespace is created by PMEM-CSI. I didn't install `ndctl` on the CoreOS worker nodes, and I had some trouble installing `ndctl` in privileged container(I will try again later). I only create the VG with name `ndbus0region0fsdax` and added `--addtag fsdax` flag as mentioned in GCE example.

**Comment by okartau** [comment]
Seems my wording "in current case" created some misunderstanding, I meant to ask about original case of that issue, not the specific test @pohly performed (I fully understood namespace was created manually there).
If namespace was created by PMEM-CSI, it for sure has been created in fsdax mode as PMEM-CSI can not do otherwise.
In that case I suspect similar effect as in #23: namespace supports DAX mode, but this support does not propagate to the LV volume, which remains without dax support and causes XFS failure.
But why this case can not be repeated in direct mode? it should not need ndctl on nodes (although that is useful for debugging)

**Comment by pohly** [comment]
> In my case with OpenShift, the namespace is created by PMEM-CSI. I didn't install ndctl on the CoreOS worker nodes, and I had some trouble installing ndctl in privileged container(I will try again later). I only create the VG with name ndbus0region0fsdax and added --addtag fsdax flag as mentioned in GCE example.

If the namespace really was created by PMEM-CSI, then why did it not also create the volume group? Something doesn't add up here.

I think what really happened here is that the PMEM device showed up as one "raw" namespace when the kernel first booted up. That namespace then was ignored by the PMEM-CSI driver. After creating the VG with the expected name, PMEM-CSI started to manage it without validating that the underlying namespace actually is in "fsdax" mode which then leads to the "not pmem" error when later some application actually asks for DAX.

What I mentioned in https://github.com/intel/pmem-csi/issues/884#issuecomment-781892567 should address this. If it does, then we discuss whether we should add a feature to PMEM-CSI where it also takes over raw namespaces. So far, our position has been that preparing fsdax namespaces is something that the cluster admin needs to take care of when deploying nodes. The reason for that is that this is a one-time operation which is potentially destructive and may depend a lot on the hardware, so it's better to put it into hardware specific deployment scripts.

Regarding running ndctl: `kubectl exec -ti pmem-csi-intel-com-node-XXX -c pmem-driver -- /bin/bash` should allow running `ndctl`. It might be affected by https://github.com/intel/pmem-csi/issues/786, though, for which we haven't merged the workaround (https://github.com/intel/pmem-csi/pull/867) because we have not been able to verify it beforehand.

**Comment by Tianyang-Zhang** [comment]
> What I mentioned in #884 (comment) should address this.

I tried this on a CentOS 8 vSphere VM and I can successfully create a namespace with fsdax mode and mount it with `-o dax`:
```
mount | grep pmem0
/dev/pmem0 on /mnt/pmem0 type xfs (rw,relatime,seclabel,attr2,dax,inode64,sunit=4096,swidth=4096,noquota)
```

Also, with your suggestion attach to `pmem-csi-intel-com-node-xxxxx` pod, I can also create a namespace with fsdax mode and mount with `-o dax` with in the `pmem-driver` container. The dax options shows in the mount option correctly:
```
root@pmem-csi-intel-com-node-2ndnm:/# ndctl list
[
  {
    "dev":"namespace0.0",
    "mode":"fsdax",
    "map":"dev",
    "size":33820770304,
    "uuid":"6bb105e7-6e83-4afa-8956-04dcd20560f0",
    "blockdev":"pmem0"
  }
]

root@pmem-csi-intel-com-node-2ndnm:/# mkfs.xfs -f /dev/pmem0
meta-data=/dev/pmem0             isize=512    agcount=4, agsize=2064256 blks
         =                       sectsz=4096  attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=1, rmapbt=0
         =                       reflink=0
data     =                       bsize=4096   blocks=8257024, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=4031, version=2
         =                       sectsz=4096  sunit=1 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0

root@pmem-csi-intel-com-node-2ndnm:/# mount -o dax /dev/pmem0 /mnt/pmem-test/

root@pmem-csi-intel-com-node-2ndnm:/# mount | grep pmem0
/dev/pmem0 on /mnt/pmem-test type xfs (rw,relatime,seclabel,attr2,dax,inode64,noquota)
```

I think your assumption makes sense, it seems PMEM-CSI directly managing the `raw` namespace. But the weird thing is that I tried the same before on a single node `k8s` cluster on top of vSphere VM(CentOS 8 worker node), and everything worked well(I don't think I manually force-create a namespace on that, only VG is created manually).

**Comment by Tianyang-Zhang** [comment]
Good news is that after change the namepsace to `fsdax` in all nodes through `pmem-csi-intel-com-node-xxx` containers, I'm finally able to get rid of the `not pmem` error. PMEM-CSI is able to correctly provision PMEM for my pods through persistentVolumeClaim and CSI Ephemeral volume in this case.

**Comment by pohly** [comment]
So how do we take this further?

Can you imagine that your customers will be able to run `ndctl` to change the namespace mode before bringing up PMEM-CSI?

Or do we need to some kind of support for this scenario in PMEM-CSI itself? What could work is this:
- in LVM mode (and only in that mode!), check for raw namespaces that are not in use
- convert those to fsdax
- enable LVM on top of it

This is a change of behavior that we probably have to make opt-in via a new driver parameter.

The "not in use" part is a bit tricky. Is it possible to convert with `ndctl create-namespace -e namespace0.0 --mode fsdax`, i.e. without `-f` = `--force`? I'm not comfortable with running something that requires explicit confirmation via that flag.

**Comment by Tianyang-Zhang** [comment]
It would be great if PMEM-CSI can handle this. It's kind of inconvenient for customers to write a script to attach to every PMEM-CSI pod, reconfigure the namespace mode, and re-deploy PMEM-CSI.

>  Is it possible to convert with `ndctl create-namespace -e namespace0.0` --mode fsdax

I deleted all PMEM-CSI and user pods in the cluster and tried it without `-f`, `ndctl` returns `Device or resource busy error`:
```
root@pmem-csi-intel-com-node-hz6xg:/# ndctl create-namespace -e namespace0.0 --mode fsdax
  Error: namespace0.0 is active, specify --force for re-configuration

failed to reconfigure namespace: Device or resource busy
```

**Comment by pohly** [comment]
Was that with a "fresh" VM or one where the namespace had already been used?

**Comment by Tianyang-Zhang** [comment]
> Was that with a "fresh" VM or one where the namespace had already been used?

I was not `fresh`, the namespace had been used.

**Comment by Tianyang-Zhang** [comment]
I tried in a centos:8 container in a fresh OpenShift Cluster, the `-f` flag is still needed to reconfigure the namespace mode.

**Comment by pohly** [comment]
I have two proposals.

The first one is to add a new `--force-convert-raw-namespaces` parameter. When the driver starts up in LVM mode, it will then first do the `ndctl create-namespace -e xxx --mode fsdax --force` for all namespaces `xxx` which are currently in raw mode. Then it creates the volume group like it normally does. The downside is that it will do that on all nodes. We have to change our deployment files and operator API to make this behavior opt-in.

The second proposal is to use a `pmem-csi.intel.com/convert-raw-namespaces=force` label for nodes to trigger this behavior. We define a DaemonSet which runs PMEM-CSI on all nodes *with* that label. When it runs, it does:
- change the namespaces
- remove the `pmem-csi.intel.com/convert-raw-namespaces=force` label
- add the labels that the "normal" DaemonSet expects
- terminates

Once that is done, the normal DaemonSet starts the pod on the converted node.

I like the second proposal a bit better. We can unconditionally enable that second DaemonSet and it won't do anything by default. Starting up the driver will take a bit longer the first time, but IMHO that's okay.

**Comment by pohly** [comment]
A third proposal is to use just one DaemonSet (i.e. as it is now) and add code in PMEM-CSI which checks node labels. If `pmem-csi.intel.com/convert-raw-namespaces=force` is set when it starts, then this enables the namespace conversion in LVM mode.

The downside of this is worse integration with Node Feature Discovery (NFD): if PMEM-CSI is configured to run on nodes with the `feature.node.kubernetes.io/memory-nv.dax=true` label, then it won't even start on those nodes which currently only have a raw namespace.

**Comment by Tianyang-Zhang** [comment]
For our current usage, all three methods sound good to me. I also like the second proposal better. It will be you and your team's call to decide how to do it. Thanks for keep tracking this issue!

**Comment by avalluri** [comment]
> The second proposal is to use a `pmem-csi.intel.com/convert-raw-namespaces=force` label for nodes to trigger this behavior. We define a DaemonSet which runs PMEM-CSI on all nodes _with_ that label. When it runs, it does:
> 
> * change the namespaces
> * remove the `pmem-csi.intel.com/convert-raw-namespaces=force` label
> * add the labels that the "normal" DaemonSet expects

How is the node labeling handled in case of NFD? Do we skip this step?

**Comment by pohly** [comment]
There is no automatic node labeling by NFD for raw namespaces, at least as far as I know. Even if there is one, we don't want to use that because converting an existing raw namespace is destructive and should require an explicit opt-in. So `pmem-csi.intel.com/convert-raw-namespaces=force` has to be set manually.

The "normal" label then can be `feature.node.kubernetes.io/memory-nv.dax=true`, i.e. the same one that NFD would also set. Setting it via the PMEM-CSI conversion step is just a shortcut because we don't need to wait for NFD to notice that there is now an fsdax namespace.

**Comment by pohly** [comment]
@avalluri so do you agree that option 2 is the right way to go?

**Comment by avalluri** [comment]
> @avalluri so do you agree that option 2 is the right way to go?

Yes,

**Comment by pohly** [comment]
There's a slight usability issue with the proposed solution: it is possible to inspect the daemonset which is responsible for converting namespaces, but then finding further information about what was done isn't possible. The logs of the pods cannot be retrieved because they get deleted immediately and it's also not visible where they ran:
```
$ kubectl describe daemonsets/pmem-csi-intel-com-node-setup -n pmem-csi
Name:           pmem-csi-intel-com-node-setup
Selector:       app.kubernetes.io/instance=pmem-csi.intel.com,app.kubernetes.io/name=pmem-csi-node-setup,pmem-csi.intel.com/deployment=lvm-testing
Node-Selector:  pmem-csi.intel.com/convert-raw-namespaces=force
Labels:         app.kubernetes.io/component=node-setup
                app.kubernetes.io/instance=pmem-csi.intel.com
                app.kubernetes.io/name=pmem-csi-node-setup
                app.kubernetes.io/part-of=pmem-csi
                pmem-csi.intel.com/deployment=lvm-testing
Annotations:    deprecated.daemonset.template.generation: 1
Desired Number of Nodes Scheduled: 0
Current Number of Nodes Scheduled: 0
Number of Nodes Scheduled with Up-to-date Pods: 0
Number of Nodes Scheduled with Available Pods: 0
Number of Nodes Misscheduled: 0
Pods Status:  0 Running / 0 Waiting / 0 Succeeded / 0 Failed
Pod Template:
  Labels:           app.kubernetes.io/component=node-setup
                    app.kubernetes.io/instance=pmem-csi.intel.com
                    app.kubernetes.io/name=pmem-csi-node-setup
                    app.kubernetes.io/part-of=pmem-csi
                    pmem-csi.intel.com/deployment=lvm-testing
                    pmem-csi.intel.com/webhook=ignore
  Service Account:  pmem-csi-intel-com-node-setup
  Containers:
   pmem-driver:
    Image:      172.17.42.1:5001/pmem-csi-driver-test:canary
    Port:       <none>
    Host Port:  <none>
    Command:
      /usr/local/bin/pmem-csi-driver
      -v=3
      -logging-format=text
      -mode=force-convert-raw-namespaces
      -nodeSelector={"storage": "pmem"}
      -nodeid=$(KUBE_NODE_NAME)
    Environment:
      KUBE_NODE_NAME:         (v1:spec.nodeName)
      TERMINATION_LOG_PATH:  /tmp/termination-log
    Mounts:
      /dev from dev-dir (rw)
      /sys from sys-dir (rw)
  Volumes:
   dev-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /dev
    HostPathType:  DirectoryOrCreate
   sys-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /sys
    HostPathType:  DirectoryOrCreate
Events:
  Type    Reason            Age    From                  Message
  ----    ------            ----   ----                  -------
  Normal  SuccessfulCreate  2m20s  daemonset-controller  Created pod: pmem-csi-intel-com-node-setup-52f6k
  Normal  SuccessfulDelete  2m19s  daemonset-controller  Deleted pod: pmem-csi-intel-com-node-setup-52f6k
  Normal  SuccessfulCreate  5s     daemonset-controller  Created pod: pmem-csi-intel-com-node-setup-hb6v8
  Normal  SuccessfulDelete  4s     daemonset-controller  Deleted pod: pmem-csi-intel-com-node-setup-hb6v8
```

**Comment by pohly** [comment]
If the conversion doesn't do anything, as in this case, then it is possible to run the same command in a regular PMEM-CSI driver pod:
```
$ kubectl exec -ti -n pmem-csi pmem-csi-intel-com-node-nsfj7 -c pmem-driver -- /bin/bash
root@pmem-csi-intel-com-node-nsfj7:/# /usr/local/bin/pmem-csi-driver -v=5 -mode=force-convert-raw-namespaces -nodeSelector='{"storage": "pmem"}' -nodeid=pmem-csi-pmem-govm-worker3
I0317 15:45:34.044221      28 main.go:69] Version: v0.8.0-250-gf482cb69a
I0317 15:45:34.046261      28 convert.go:157] "raw-namespace-conversion: checking" namespace="{\"blockdev\":\"pmem0\",\"dev\":\"namespace0.0\",\"enabled\":true,\"id\":0,\"mode\":\"raw\",\"name\":\"\",\"size\":68719476736,\"uuid\":\"00000000-0000-0000-0000-000000000000\"}"
I0317 15:45:34.046339      28 convert.go:147] "raw-namespace-conversion: successful" converted=0
I0317 15:45:34.046358      28 convert.go:230] "relabel: node" patch="{\"metadata\":{\"labels\":{\"pmem-csi.intel.com/convert-raw-namespaces\": null, \"storage\": \"pmem\"}}}"
failed to run driver: relabel node pmem-csi-pmem-govm-worker3: failed to patch node: nodes "pmem-csi-pmem-govm-worker3" is forbidden: User "system:serviceaccount:pmem-csi:pmem-csi-intel-com-controller" cannot patch resource "nodes" in API group "" at the cluster scope:
PASS
coverage: 6.6% of statements in ./pkg/...
```

I need to add more logging, but at least it is possible to see it (albeit in a rather difficult way).

**Comment by pohly** [comment]
@Tianyang-Zhang : I implemented option 2 in PR https://github.com/intel/pmem-csi/pull/919.

What I don't know is whether it really works on vSphere. I was only able to test on QEMU. The PR has an E2E test that runs on that. Perhaps you can help with testing it on vSphere? Do we need to merge the PR for that first (in which case the "canary" images have the necessary support) or can you build from source?

While writing documentation, I realized that a design choice might not be suitable:
> If the conversion finds no raw namespaces on a node, it will remove
> `convert-raw-namespaces` label and exit without adding the normal
> label, i.e. the node will not be used for PMEM-CSI. This is not
> considered an error, which may be useful in clusters where it is not
> certain in advance which nodes will have a raw namespace.
> ...
> A log monitoring solution is needed to catch this output because these
> pods are short-lived and `kubectl logs` does not work for pods that
> were already deleted.

Would it be better to treat a missing raw namespace as error, i.e. exit with non-zero exit code and not remove the label? Then the pod does not get deleted and its log output can be inspected easily to figure out what it found (or rather, did not find) on the node.

**Comment by Tianyang-Zhang** [comment]
Thanks for updating @pohly ! I built the image from your branch `pohly:raw-namespace-conversion` and tested on OpenShift 4 + vSphere environment. I added the new label to each node and I can see the daemonset is created as expected: 
```
$ oc get daemonset
NAME                            DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR                                     AGE
pmem-csi-intel-com-node         3         3         0       3            0           storage=pmem                                      2s
pmem-csi-intel-com-node-setup   3         3         0       3            0           pmem-csi.intel.com/convert-raw-namespaces=force   2s
```

But the pods are not converting the raw PMEM device to fsdax mode, no matter I create VG on each node or not:
```
$ oc get pods -o wide -n pmem-csi
NAME                                  READY   STATUS             RESTARTS   AGE   IP            NODE                     NOMINATED NODE   READINESS GATES
pmem-csi-intel-com-controller-0       1/1     Running            0          20s   10.128.0.74   osc-5k68w-master-1       <none>           <none>
pmem-csi-intel-com-node-2rd8f         3/3     Running            0          20s   10.129.2.11   osc-5k68w-worker-7c4g5   <none>           <none>
pmem-csi-intel-com-node-m6sj8         3/3     Running            0          20s   10.131.0.10   osc-5k68w-worker-jvd4p   <none>           <none>
pmem-csi-intel-com-node-setup-tn6qz   0/1     CrashLoopBackOff   1          20s   10.131.0.9    osc-5k68w-worker-jvd4p   <none>           <none>
pmem-csi-intel-com-node-setup-z96p2   0/1     CrashLoopBackOff   1          20s   10.128.2.13   osc-5k68w-worker-9c42p   <none>           <none>
pmem-csi-intel-com-node-setup-zq4sx   0/1     Error              2          20s   10.129.2.10   osc-5k68w-worker-7c4g5   <none>           <none>
pmem-csi-intel-com-node-xmnc8         3/3     Running            0          20s   10.128.2.14   osc-5k68w-worker-9c42p   <none>           <none>
```

The logs shows as following:
```
oc logs -f pmem-csi-intel-com-node-setup-zq4sx
I0323 00:24:49.123326       1 main.go:69] Version: v0.8.0-266-g1b8bb713
I0323 00:24:49.125085       1 convert.go:157] "raw-namespace-conversion: checking for namespaces"
I0323 00:24:49.125517       1 convert.go:159] "raw-namespace-conversion: checking" bus="{\"dev\":\"ndbus0\",\"dimms\":[{}],\"provider\":\"ACPI.NFIT\",\"region
s\":[{}]}"
failed to run driver: Failed to set namespace uid: No such file or directory
I0323 00:24:49.125926       1 convert.go:161] "raw-namespace-conversion: checking" region="{\"available_size\":0,\"dev\":\"region0\",\"mappings\":[{}],\"max_a
vailable_extent\":0,\"namespaces\":null,\"size\":34359738368,\"type\":\"pmem\"}"
I0323 00:24:49.126067       1 convert.go:178] "raw-namespace-conversion: converting raw namespace" namespace="{\"blockdev\":\"\",\"dev\":\"namespace0.0\",\"en
abled\":false,\"id\":0,\"mode\":\"raw\",\"name\":\"\",\"size\":34359738368,\"uuid\":\"00000000-0000-0000-0000-000000000000\"}"
E0323 00:24:49.126299       1 convert.go:151] "raw-namespace-conversion: failed" err="Failed to set namespace uid: No such file or directory" converted=0
```

After the daemonset ran, the `/dev/pmem0` device is gone, and `ndctl list` doesn't show anything. I need to reboot all host to recover. 

For your question
> Would it be better to treat a missing raw namespace as error, i.e. exit with non-zero exit code and not remove the label

IMHO, maybe it is better to act differently in 2 cases:
1. No raw namespace and no fsdax namespace were found. In this case, we may treat it as an error and not remove the label.
2. No raw namespace found, but found fsdax namespace(s). In this case, I think we can treat it as no error and remove the label.

What do you think?

**Comment by pohly** [comment]
On the bright side, at least you were able to retrieve logs in this case :sweat_smile: 

They just don't help a lot, unfortunately. Somehow my implementation of `ndctl create-namespace --convert` must be different enough that it worked for me under QEMU, but fails under vSphere. Replicating the exact logic of that particular operation was not straight-forward. I wonder whether it is worthwhile to even try it - calling the external `ndctl` would be simpler. The downside of that is that it provides less control (in theory, at least) and we introduce a dependency on the tool.

Regarding error handling: I like your proposal. The first case makes it easy to debug when something unexpected happened. The second case covers idempotency if namespace conversion worked, but then (intermittently) communication with the API server failed.

**Comment by pohly** [comment]
@avalluri what do you think about calling the ndctl binary in this case?

In general, what was the rationale for writing Go bindings for libndctl instead of just calling the binary and parsing the JSON output?

**Comment by okartau** [comment]
> In general, what was the rationale for writing Go bindings for libndctl instead of just calling the binary and parsing the JSON output?

I recall we considered:
- library call being lower overhead/faster than running binary (which ends up calling the same library anyway).
- result as return value is easier to parse than writing (new) parser for text-form output which tends to change more often that library API

If we have case where calling binary is better, then of course this can be reconsidered.

**Comment by avalluri** [comment]
> @avalluri what do you think about calling the ndctl binary in this case?
>
 `
convert.go:151] "raw-namespace-conversion: failed" err="Failed to set namespace uid: No such file or directory" converted=0
`
By seeing the error, I would propose to try to add a check `if ns.Type() != IoNamespace{ }` while calling `ns.SetUUID()`.

If that won't help, maybe better to use ndctl binary.

> In general, what was the rationale for writing Go bindings for libndctl instead of just calling the binary and parsing the JSON output?

At the initial prototyping stage, we used ndctl binary directly. Later when we were aware of libndctl, just moved to use that assuming that we get more flexibility/control.

**Comment by pohly** [comment]
https://github.com/pmem/ndctl/blob/ea014c0c9ec8d0ef945d072dcc52b306c7a686f9/ndctl/namespace.c#L493-L503 has such a check. But can we really skip that SetUUID, SetSize and SetAltName in this case? In particular the name is important. Without it, the LVM mode will not set up the volume group.

@Tianyang-Zhang : may I ask you to test whether at least the conversion now works? I suspect I'll have to add volume group setup to the conversion step instead of relying on the normal driver startup for that, but I can add that once I know that the conversion to fsdax works.

**Comment by Tianyang-Zhang** [comment]
Sure. I rebuild the image and did the test again. This time I get one pod end up in `complete` state but the other 2 still 
in `CrashLoopBackOff` state:
```
oc get pods -n pmem-csi -o wide
NAME                                  READY   STATUS             RESTARTS   AGE   IP            NODE                     NOMINATED NODE   READINESS GATES
pmem-csi-intel-com-controller-0       1/1     Running            0          20s   10.128.0.79   osc-5k68w-master-1       <none>           <none>
pmem-csi-intel-com-node-9zp78         3/3     Running            0          20s   10.128.3.85   osc-5k68w-worker-9c42p   <none>           <none>
pmem-csi-intel-com-node-hg85f         3/3     Running            0          20s   10.129.2.83   osc-5k68w-worker-7c4g5   <none>           <none>
pmem-csi-intel-com-node-setup-5wfdj   0/1     Completed          1          20s   10.131.0.57   osc-5k68w-worker-jvd4p   <none>           <none>
pmem-csi-intel-com-node-setup-ml8rb   0/1     CrashLoopBackOff   1          20s   10.129.2.84   osc-5k68w-worker-7c4g5   <none>           <none>
pmem-csi-intel-com-node-setup-w2bfw   0/1     CrashLoopBackOff   1          20s   10.128.3.86   osc-5k68w-worker-9c42p   <none>           <none>
pmem-csi-intel-com-node-tz56k         3/3     Running            0          20s   10.131.0.56   osc-5k68w-worker-jvd4p   <none>           <none>
```

Then the `complete` pod also turns to `CrashLoopBackOff` state. The log of three pods are the same:
```
oc logs pmem-csi-intel-com-node-setup-5wfdj
I0326 03:03:51.754546       1 main.go:69] Version: v0.9.0-19-g52dcd362-dirty
I0326 03:03:51.756486       1 convert.go:157] "raw-namespace-conversion: checking for namespaces"
I0326 03:03:51.756953       1 convert.go:159] "raw-namespace-conversion: checking" bus="{\"dev\":\"ndbus0\",\"dimms\":[{}],\"provider\":\"ACPI.NFIT\",\"regions\":[{}]}"
failed to run driver: set fsdax mode: Failed to set enforce mode: No such device or address
PASS
coverage: 8.0% of statements in ./pkg/...
I0326 03:03:51.757406       1 convert.go:161] "raw-namespace-conversion: checking" region="{\"available_size\":0,\"dev\":\"region0\",\"mappings\":[{}],\"max_available_extent\":0,\"namespaces\":null,\"size\":34359738368,\"type\":\"pmem\"}"
I0326 03:03:51.757554       1 convert.go:178] "raw-namespace-conversion: converting raw namespace" namespace="{\"blockdev\":\"\",\"dev\":\"namespace0.0\",\"enabled\":false,\"id\":0,\"mode\":\"raw\",\"name\":\"\",\"size\":34359738368,\"uuid\":\"00000000-0000-0000-0000-000000000000\"}"
E0326 03:03:51.757672       1 convert.go:151] "raw-namespace-conversion: failed" err="set fsdax mode: Failed to set enforce mode: No such device or address" converted=0
```

The PMEM device is gone as same as the last build, `oc exec -it pmem-csi-intel-com-node-tz56k -c pmem-driver -- ndctl list` returns nothing.

Then I rebooted all nodes and ran again. This time I got:
```
oc get pods -n pmem-csi -o wide
NAME                                  READY   STATUS      RESTARTS   AGE   IP            NODE                     NOMINATED NODE   READINESS GATES
pmem-csi-intel-com-controller-0       1/1     Running     0          25s   10.128.0.80   osc-5k68w-master-1       <none>           <none>
pmem-csi-intel-com-node-8dvcn         3/3     Running     0          25s   10.128.2.14   osc-5k68w-worker-9c42p   <none>           <none>
pmem-csi-intel-com-node-fdjtj         3/3     Running     0          25s   10.131.0.9    osc-5k68w-worker-jvd4p   <none>           <none>
pmem-csi-intel-com-node-kwznm         3/3     Running     0          25s   10.129.2.9    osc-5k68w-worker-7c4g5   <none>           <none>
pmem-csi-intel-com-node-setup-684f4   0/1     Completed   2          25s   10.131.0.10   osc-5k68w-worker-jvd4p   <none>           <none>
pmem-csi-intel-com-node-setup-njbkx   0/1     Completed   2          25s   10.129.2.10   osc-5k68w-worker-7c4g5   <none>           <none>
pmem-csi-intel-com-node-setup-pql6h   0/1     Completed   2          25s   10.128.2.13   osc-5k68w-worker-9c42p   <none>           <none>
```

But the logs seems still showing error:
```
oc logs pmem-csi-intel-com-node-setup-684f4
I0326 03:08:00.396148       1 main.go:69] Version: v0.9.0-19-g52dcd362-dirty
I0326 03:08:00.398067       1 convert.go:157] "raw-namespace-conversion: checking for namespaces"
I0326 03:08:00.398499       1 convert.go:159] "raw-namespace-conversion: checking" bus="{\"dev\":\"ndbus0\",\"dimms\":[{}],\"provider\":\"ACPI.NFIT\",\"regions\":[{}]}"
I0326 03:08:00.398869       1 convert.go:161] "raw-namespace-conversion: checking" region="{\"available_size\":0,\"dev\":\"region0\",\"mappings\":[{}],\"max_available_extent\":0,\"namespaces\":null,\"size\":34359738368,\"type\":\"pmem\"}"
failed to run driver: set fsdax mode: Failed to set enforce mode: No such device or address
PASS
coverage: 8.0% of statements in ./pkg/...
I0326 03:08:00.399509       1 convert.go:178] "raw-namespace-conversion: converting raw namespace" namespace="{\"blockdev\":\"\",\"dev\":\"namespace0.0\",\"enabled\":false,\"id\":0,\"mode\":\"raw\",\"name\":\"\",\"size\":34359738368,\"uuid\":\"00000000-0000-0000-0000-000000000000\"}"
E0326 03:08:00.399711       1 convert.go:151] "raw-namespace-conversion: failed" err="set fsdax mode: Failed to set enforce mode: No such device or address" converted=0
```

And `ndctl list` still returns nothing: 
```
oc exec -it pmem-csi-intel-com-node-8dvcn -c pmem-driver -- /bin/bash
root@pmem-csi-intel-com-node-8dvcn:/# ls -l /dev/pmem*
ls: cannot access '/dev/pmem*': No such file or directory
root@pmem-csi-intel-com-node-8dvcn:/# ndctl list
root@pmem-csi-intel-com-node-8dvcn:/#
```

**Comment by pohly** [comment]
Okay, I'll switch to calling the `ndctl` binary. That seems easier than trying to reimplement it.

However, there is one open question: can we set the `pmem-csi` name under vSphere?

@Tianyang-Zhang can you perhaps run `ndctl create-namespace -e namespace0.0 --mode fsdax --name pmem-csi --force` on a machine which has a namespace0.0 in raw mode and then verify with `ndctl list -NR` that the name is set?

I'm starting to suspect that the name will be silently dropped by `ndctl`. In that case there's further work needed on setting up the volume group.

**Comment by Tianyang-Zhang** [comment]
It seems the name is dropped by `ndctl`:
```
[root@pmem-fsdax-config-8b5wc /]# ndctl create-namespace -e namespace0.0 --mode fsdax --name pmem-csi --force
{
  "dev":"namespace0.0",
  "mode":"fsdax",
  "map":"dev",
  "size":"31.50 GiB (33.82 GB)",
  "uuid":"5cd0222e-64f0-4f5a-b9a0-36d418307bb4",
  "sector_size":512,
  "align":2097152,
  "blockdev":"pmem0"
}
```

The output of `ndctl list -NR`
```
[root@pmem-fsdax-config-8b5wc /]# ndctl list -NR
{
  "regions":[
    {
      "dev":"region0",
      "size":34359738368,
      "available_size":0,
      "max_available_extent":0,
      "type":"pmem",
      "persistence_domain":"memory_controller",
      "namespaces":[
        {
          "dev":"namespace0.0",
          "mode":"fsdax",
          "map":"dev",
          "size":33820770304,
          "uuid":"5cd0222e-64f0-4f5a-b9a0-36d418307bb4",
          "sector_size":512,
          "align":2097152,
          "blockdev":"pmem0"
        }
      ]
    }
  ]
}
```

**Comment by pohly** [comment]
Thanks for the confirmation. I'll work around that.

**Comment by pohly** [comment]
> IMHO, maybe it is better to act differently in 2 cases:
> 
> -    No raw namespace and no fsdax namespace were found. In this case, we may treat it as an error and not remove the label.
>  -  No raw namespace found, but found fsdax namespace(s). In this case, I think we can treat it as no error and remove the label.

I've implemented that, and the idempotency aspect of it worked rather well even when the setup pod was started twice: it first did its work and must have removed labels, but then it was restarted by the daemonset controller before that noticed the label changes. The second run then didn't change anything and eventually the driver pod started up.

Having said that, the pod shouldn't have been started twice. It seems related to the operator and how it updated the DaemonSet (unnecessarily?):

```
pmem-csi-operator-5869fcc4b8-hgdth/pmem-csi-operator@pmem..ker1: I0406 18:15:26.412498       1 deployment_controller.go:126] "UPDATED" object="default/second-pmem-csi-intel-com-node-setup <*v1.DaemonSet>"
pmem-csi-operator-5869fcc4b8-hgdth/pmem-csi-operator@pmem..ker1: I0406 18:15:26.412976       1 deployment_controller.go:466] "newDeployment: load controller TLS secret" secret="default/second-pmem-csi-intel-com-controller-secret" CAlength=1086
pmem-csi-operator-5869fcc4b8-hgdth/pmem-csi-operator@pmem..ker1: I0406 18:15:26.413093       1 controller_driver.go:627] "deployment/event: start" object="default/second-pmem-csi-intel-com-node-setup <*v1.DaemonSet>" type="*v1.DaemonSet"
pmem-csi-operator-5869fcc4b8-hgdth/pmem-csi-operator@pmem..ker1: I0406 18:15:26.413289       1 controller_driver.go:641] "deployment/event: redeploying" name="node setup driver" object="default/second-pmem-csi-intel-com-node-setup <*v1.DaemonSet>"
pmem-csi-operator-5869fcc4b8-hgdth/pmem-csi-operator@pmem..ker1: I0406 18:15:26.413454       1 controller_driver.go:194] "redeploy/getSubObject: get" object="default/second-pmem-csi-intel-com-node-setup <*v1.DaemonSet>"
pmem-csi-operator-5869fcc4b8-hgdth/pmem-csi-operator@pmem..ker1: I0406 18:15:26.415331       1 controller_driver.go:280] "redeploy: update" object="default/second-pmem-csi-intel-com-node-setup"
```

The status of `second-pmem-csi-intel-com-node-setup` changed as part of the first pod termination, but that should have caused an updated for that DaemonSet.

This was for `make test_e2e TEST_E2E_FOCUS=operator.*lvm-production.*raw-conversion.*works` on my branch.

**Comment by pohly** [comment]
It's not clear to me from the log messages at -v=5 whether the operator actually does something:
```
pmem-csi-operator-5869fcc4b8-49qmm/pmem-csi-operator@pmem..ker2: I0406 18:53:19.837132       1 controller_driver.go:280] "redeploy: update" object="default/second-pmem-csi-intel-com-node"
pmem-csi-operator-5869fcc4b8-49qmm/pmem-csi-operator@pmem..ker2: I0406 18:53:19.851437       1 deployment_controller.go:67] "UPDATED" object="second.pmem-csi.intel.com <*v1beta1.PmemCSIDeployment>" generation=1
```

That's for `TEST_OPERATOR_LOGLEVEL=5` -> `-vmodule=*=5`. Why is the `-vmodule` parameter used instead of the simpler `-v`?

**Comment by pohly** [comment]
Yes, line 280 does apply a patch, but without the patch it remains unclear what change it is making. I'll add that information.

**Comment by pohly** [comment]
The patch is a lot larger than I expected:
```
pmem-csi-operator-7d8f55f866-pc2ht/pmem-csi-operator@pmem..ker3: I0406 20:08:35.694748       1 controller_driver.go:280] "redeploy: update" object="default/second-pmem-csi-intel-com-node-setup" patch="{\"spec\":{\"revisionHistoryLimit\":null,\"template\":{\"spec\":{\"containers\":[{\"command\":[\"/usr/local/bin/pmem-csi-driver\",\"-v=3\",\"-logging-format=text\",\"-mode=force-convert-raw-namespaces\",\"-nodeSelector={\\\"feature.node.kubernetes.io/memory-nv.dax\\\":\\\"true\\\"}\",\"-nodeid=$(KUBE_NODE_NAME)\"],\"env\":[{\"name\":\"KUBE_NODE_NAME\",\"valueFrom\":{\"fieldRef\":{\"apiVersion\":\"v1\",\"fieldPath\":\"spec.nodeName\"}}},{\"name\":\"TERMINATION_LOG_PATH\",\"value\":\"/tmp/termination-log\"}],\"image\":\"172.17.42.1:5001/pmem-csi-driver:canary\",\"imagePullPolicy\":\"IfNotPresent\",\"name\":\"pmem-driver\",\"resources\":{},\"securityContext\":{\"privileged\":true,\"runAsUser\":0},\"terminationMessagePath\":\"/tmp/termination-log\",\"volumeMounts\":[{\"mountPath\":\"/dev\",\"name\":\"dev-dir\"},{\"mountPath\":\"/sys\",\"name\":\"sys-dir\"},{\"mountPath\":\"/host-sys\",\"name\":\"sys-dir\"}]}],\"dnsPolicy\":null,\"restartPolicy\":null,\"schedulerName\":null,\"securityContext\":null,\"serviceAccount\":null,\"terminationGracePeriodSeconds\":null}},\"updateStrategy\":{\"rollingUpdate\":null,\"type\":null}}}"
```

@avalluri is that normal? Shouldn't it only patch the fields that really need to be changed? Is there ever a situation when the `if string(data) != "{}" && len(data) >= 0` causes an update to be skipped?

**Comment by avalluri** [comment]
> That's for `TEST_OPERATOR_LOGLEVEL=5` -> `-vmodule=*=5`. Why is the `-vmodule` parameter used instead of the simpler `-v`?

I had to use `-vmodule` when setting `-v` did not work. But I realized that there is a typo in sed regex. Fixing that could remove `-vmodule` usage. I figured this out while working on version skew tests.

**Comment by pohly** [comment]
I also found and fixed the `-v` sed regex issue. I have that together with `-vmodule` removal in my branch.

**Comment by avalluri** [comment]
> @avalluri is that normal? Shouldn't it only patch the fields that really need to be changed?

The diff supposed to contain only changed fields. But I am not sure what changed.

>  Is there ever a situation when the `if string(data) != "{}" && len(data) >= 0` causes an update to be skipped?

I remember the patch data used to  have '{}' for empty diff, that means it used to skip those cases. But I am not sure what changed to give that long diff. I will check.

**Comment by pohly** [comment]
Restarting the "node-setup" pod seems related to the operator behavior. I don't see that when testing without it.

**Comment by pohly** [comment]
I think I have addressed all opens. The revised version that I just pushed creates the volume group during conversion without relying on the name and the error handling is different.

The operator patch issue can be investigated in parallel.

@Tianyang-Zhang can you perhaps give it one more try?

@avalluri can you review?

**Comment by pohly** [comment]
> The operator patch issue can be investigated in parallel.

It's worth pointing out that it no longer causes the pod to restart: I changed the exit handling of the conversion pod so that it keeps running until terminated by Kubernetes, i.e. there is no longer a race between "pod terminates itself" and "Kubernetes restarts pod". That seems to help.

**Comment by Tianyang-Zhang** [comment]
I tried the newest changes on our `vSphere + OpenShift` environment and it works! The raw namespaces are converted to `fsdax`, volume groups is created on each node, and I am able to claim PMEM through CSI ephemeral volume.

I also ran this on the cluster that has all node already has `fsdax` mode PMEM, and I don't see any issue.

Thanks again for keep updating and resolving this issue!
